packages/cojson-transport-ws/src/index.ts

```diff
@@ -29,72 +47,73 @@ export function createWebSocketPeer(options: {
-                    incoming,
-                    Either.left(
-                        new DisconnectedError(`${event.code}: ${event.reason}`),
-                    ),
+        yield* addEventListener(ws, "close", (event) =>
+            Queue.offer(
+                incoming,
+                Either.left(
+                    new DisconnectedError(`${event.code}: ${event.reason}`),
                 ),
-            ).then((e) => {
-                if (Exit.isFailure(e) && !Exit.isInterrupted(e)) {
-                    console.warn("Failed closing ws", e);
-                }
-            });
-        });
+            ).pipe(
+                Effect.onExit((e) => {
+                    if (Exit.isFailure(e) || Exit.isInterrupted(e)) {
+                        return Console.warn("Failed closing ws", e);
+                    }
+                    return Effect.void;
+                }),
+            ),
+        );
 
-        let pingTimeout: ReturnType<typeof setTimeout> | undefined;
+        let pingTimeout: Fiber.Fiber<void> | undefined;
 
-        ws.addEventListener("message", (event) => {
-            const msg = JSON.parse(event.data as string);
+        yield* addEventListener(ws, "message", (event) =>
+            Effect.gen(function* () {
+                const msg = yield* Effect.sync(() =>
+                    JSON.parse(event.data as string),
+                );
 
-            if (pingTimeout) {
-                clearTimeout(pingTimeout);
-            }
+                if (pingTimeout) {
+                    yield* Fiber.interrupt(pingTimeout);
+                }
 
-            pingTimeout = setTimeout(() => {
-                console.debug("Ping timeout");
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.left(new PingTimeoutError())),
-                );
-                try {
-                    ws.close();
-                } catch (e) {
-                    console.error(
-                        "Error while trying to close ws on ping timeout",
-                        e,
+                pingTimeout = yield* Effect.gen(function* () {
+                    yield* Effect.sleep("2500 millis");
+                    yield* Console.warn("Ping timeout");
+                    yield* Queue.offer(
+                        incoming,
+                        Either.left(new PingTimeoutError()),
+                    ).pipe(Effect.forkScoped);
+                    try {
+                        ws.close();
+                    } catch (e) {
+                        yield* Console.error(
+                            "Error while trying to close ws on ping timeout",
+                            e,
+                        );
+                    }
+                }).pipe(Effect.scoped, Effect.forkDaemon);
+
+                if (msg?.type !== "ping") {
+                    yield* Queue.offer(incoming, Either.right(msg)).pipe(
+                        Effect.forkDaemon,
                     );
```

Comment:
you did Effect.runPromise so I assumed you wanted to fork this

Reply:
It doesn't make sense, offering is a sync action. I believe that this is related to a situation where the `Queue` is being shutdown, if `offer`ed then it would **interrupt** the effect. In other words, it's either not required or "hiding" a bug